### DISCUSSION TITLE
Multiple Transcoding

### DISCUSSION THREAD

**Discussion by lsirikh** [bluenviron/mediamtx#3242] [Questions]
### Question

I'm currently building and using an RTSP media server using MediaMTX. To control MediaMTX, I'm using the API to register, delete, and change things.

The environmental information is as follows.
OS: Window10 pro
MediaMTX version: MediaMTX v1.3.1
Purpose of Use: Media Servers for VMS Servers 
Development Language: C# (WPF)

The issues I'm currently facing are as follows.

The method we were going to develop previously was that VMS received all profiles streamed from the IP camera to adjust the resolution according to Grid View.

For example), from the HikVision camera 192.168.1.119 IP, all three profiles (Main, Second, Third) rtsp profile addresses were registered through the API with MediaMTX. We configured the logic for the VMS client to operate by streaming the first profile address of that camera when Grid View is 2x2, streaming the second profile address when 4x4, and streaming the third profile address when 5x5. However, the problem was that it took a lot of load on the network bandwidth because each camera imported all the profiles and registered them with the MediaMTX. So we ended up approaching it with the following solution.

1. Register one main profile on MediaMtx from one camera (e.g. rtssp://user:pass@192.168.202.150:8554/cam119_Main)
path configuration
```
{
			"name": "118_Main",
			"source": "rtsp://user:pass@192.168.1.118:554/cam/realmonitor?channel=1&subtype=0&unicast=true&proto=Onvif",
			"sourceFingerprint": "",
			"sourceOnDemand": false,
			"sourceOnDemandStartTimeout": "10s",
			"sourceOnDemandCloseAfter": "10s",
			"maxReaders": 0,
			"srtReadPassphrase": "",
			"fallback": "",
			"record": false,
			"recordPath": "./recordings/%path/%Y-%m-%d_%H-%M-%S-%f",
			"recordFormat": "fmp4",
			"recordPartDuration": "100ms",
			"recordSegmentDuration": "1h0m0s",
			"recordDeleteAfter": "24h0m0s",
			"publishUser": "",
			"publishPass": "",
			"publishIPs": [],
			"readUser": "user",
			"readPass": "pass",
			"readIPs": [],
			"overridePublisher": true,
			"srtPublishPassphrase": "",
			"rtspTransport": "automatic",
			"rtspAnyPort": false,
			"rtspRangeType": "",
			"rtspRangeStart": "",
			"sourceRedirect": "",
			"rpiCameraCamID": 0,
			"rpiCameraWidth": 1920,
			"rpiCameraHeight": 1080,
			"rpiCameraHFlip": false,
			"rpiCameraVFlip": false,
			"rpiCameraBrightness": 0,
			"rpiCameraContrast": 1,
			"rpiCameraSaturation": 1,
			"rpiCameraSharpness": 1,
			"rpiCameraExposure": "normal",
			"rpiCameraAWB": "auto",
			"rpiCameraDenoise": "off",
			"rpiCameraShutter": 0,
			"rpiCameraMetering": "centre",
			"rpiCameraGain": 0,
			"rpiCameraEV": 0,
			"rpiCameraROI": "",
			"rpiCameraHDR": false,
			"rpiCameraTuningFile": "",
			"rpiCameraMode": "",
			"rpiCameraFPS": 30,
			"rpiCameraIDRPeriod": 60,
			"rpiCameraBitrate": 1000000,
			"rpiCameraProfile": "main",
			"rpiCameraLevel": "4.1",
			"rpiCameraAfMode": "continuous",
			"rpiCameraAfRange": "normal",
			"rpiCameraAfSpeed": "normal",
			"rpiCameraLensPosition": 0,
			"rpiCameraAfWindow": "",
			"rpiCameraTextOverlayEnable": false,
			"rpiCameraTextOverlay": "%Y-%m-%d %H:%M:%S - Sensorway",
			"runOnInit": "",
			"runOnInitRestart": false,
			"runOnDemand": "",
			"runOnDemandRestart": false,
			"runOnDemandStartTimeout": "10s",
			"runOnDemandCloseAfter": "10s",
			"runOnUnDemand": "",
			"runOnReady": "",
			"runOnReadyRestart": false,
			"runOnNotReady": "",
			"runOnRead": "",
			"runOnReadRestart": false,
			"runOnUnread": "",
			"runOnRecordSegmentCreate": "",
			"runOnRecordSegmentComplete": ""
		}
```

2. Register a new low-resolution Rtsp channel using the RTSP channel address of the registered MediaMTx.
(예: "runOnInit": "ffmpeg -re -i rtsp://user:pass@192.168.202.150:8554/118_Main -vf \"scale=640:480:flags=lanczos\" -c:v libx264 -b:v 1000k -c:a copy -preset veryfast -tune zerolatency -f rtsp rtsp://192.168.202.150:8554/118_Sub")
path configuration
```
{
			"name": "118_Sub",
			"source": "publisher",
			"sourceFingerprint": "",
			"sourceOnDemand": false,
			"sourceOnDemandStartTimeout": "10s",
			"sourceOnDemandCloseAfter": "10s",
			"maxReaders": 0,
			"srtReadPassphrase": "",
			"fallback": "",
			"record": false,
			"recordPath": "./recordings/%path/%Y-%m-%d_%H-%M-%S-%f",
			"recordFormat": "fmp4",
			"recordPartDuration": "100ms",
			"recordSegmentDuration": "1h0m0s",
			"recordDeleteAfter": "24h0m0s",
			"publishUser": "",
			"publishPass": "",
			"publishIPs": [],
			"readUser": "user",
			"readPass": "pass",
			"readIPs": [],
			"overridePublisher": true,
			"srtPublishPassphrase": "",
			"rtspTransport": "automatic",
			"rtspAnyPort": false,
			"rtspRangeType": "",
			"rtspRangeStart": "",
			"sourceRedirect": "",
			"rpiCameraCamID": 0,
			"rpiCameraWidth": 1920,
			"rpiCameraHeight": 1080,
			"rpiCameraHFlip": false,
			"rpiCameraVFlip": false,
			"rpiCameraBrightness": 0,
			"rpiCameraContrast": 1,
			"rpiCameraSaturation": 1,
			"rpiCameraSharpness": 1,
			"rpiCameraExposure": "normal",
			"rpiCameraAWB": "auto",
			"rpiCameraDenoise": "off",
			"rpiCameraShutter": 0,
			"rpiCameraMetering": "centre",
			"rpiCameraGain": 0,
			"rpiCameraEV": 0,
			"rpiCameraROI": "",
			"rpiCameraHDR": false,
			"rpiCameraTuningFile": "",
			"rpiCameraMode": "",
			"rpiCameraFPS": 30,
			"rpiCameraIDRPeriod": 60,
			"rpiCameraBitrate": 1000000,
			"rpiCameraProfile": "main",
			"rpiCameraLevel": "4.1",
			"rpiCameraAfMode": "continuous",
			"rpiCameraAfRange": "normal",
			"rpiCameraAfSpeed": "normal",
			"rpiCameraLensPosition": 0,
			"rpiCameraAfWindow": "",
			"rpiCameraTextOverlayEnable": false,
			"rpiCameraTextOverlay": "%Y-%m-%d %H:%M:%S - Sensorway",
			"runOnInit": "ffmpeg -re -i rtsp://user:pass@192.168.202.150:8554/118_Main -vf \"scale=640:480:flags=lanczos\" -c:v libx264 -b:v 1000k -c:a copy -preset veryfast -tune zerolatency -f rtsp rtsp://192.168.202.150:8554/118_Sub",
			"runOnInitRestart": false,
			"runOnDemand": "",
			"runOnDemandRestart": false,
			"runOnDemandStartTimeout": "10s",
			"runOnDemandCloseAfter": "10s",
			"runOnUnDemand": "",
			"runOnReady": "",
			"runOnReadyRestart": false,
			"runOnNotReady": "",
			"runOnRead": "",
			"runOnReadRestart": false,
			"runOnUnread": "",
			"runOnRecordSegmentCreate": "",
			"runOnRecordSegmentComplete": ""
		}
```

The problem is that some cameras have a camera (HikVision) that has a delay of between one and two seconds of camera streaming and does not break the frame and comes out smoothly, while others have a delay of more than 10 seconds and break so badly that they do not even recognize the frame (Dahua).

I tried changing the codec of runOnInit and changing the -b:v value, but Dahua or other low-cost products had serious transcoding problems.

In my opinion, on MediaMtx, rtsp://user:pass@192.168.202.150:8554/118_Main is shown to create a pipeline, and transcoding within this pipeline, so I don't know why there are cameras that work well for each camera, and those that perform beyond use.

If anyone has solved this issue, please let me know.

The ID and off-duty number were unified to user/password for security purposes.

**Comment by aler9** [comment]
Hello, if i understood correctly you want to generate sub-streams locally rather than getting them from cameras, in order to save bandwidth, and you're having issues in generating these sub-streams with FFmpeg.

> some cameras have a camera (HikVision) that has a delay of between one and two seconds of camera streaming and does not break the frame and comes out smoothly, while others have a delay of more than 10 seconds and break so badly that they do not even recognize the frame (Dahua).

This should be simple to solve: decrease the I-frame interval inside the settings of problematic cameras. It controls the period between independent frames (I frames), that reset the screen and make artifacts disappear. If artifacts are shown for several seconds, it means that no I frames are being received within that period.

> "runOnInit": "ffmpeg -re -i rtsp://user:pass@192.168.202.150:8554/118_Main -vf "scale=640:480:flags=lanczos" -c:v libx264 -b:v 1000k -c:a copy -preset veryfast -tune zerolatency -f rtsp rtsp://192.168.202.150:8554/118_Sub"

Remove the `-tune zerolatency` setting. That is meant to drop frames in case of delays, which seems to be happening.

Furthermore, standard CPUs can handle a very limited amount of parallel H264 encodings, you should consider moving some or all encodings to the GPU. Instructions are in the FFmpeg wiki.

Let me emphasize that this issue is almost entirely FFmpeg-related. We have no influence on the FFmpeg behavior.
